### PR TITLE
[lang] Texture support 1/n (Context & Programs)

### PR DESCRIPTION
**Pull request by bobcao3**
Related issue = #5109

### CODE REVIEW

File: taichi/program/context.h [hunk 1]
```diff
@@ -32,6 +34,8 @@ struct RuntimeContext {
   uint64 array_runtime_sizes[taichi_max_num_args_total]{0};
   // `is_device_allocations` is true iff i-th arg is a `DeviceAllocation*`.
   bool is_device_allocations[taichi_max_num_args_total]{false};
```
Comment (k-ye) [inline taichi/program/context.h:36]: can we remove `is_device_allocation` now? (Maybe we need to add something like `DevAllocType::kInvalid`.

File: taichi/program/context.h [hunk 2]
```diff
@@ -14,6 +14,8 @@ struct DeviceAllocation;
 // pointer to the LLVMRuntime struct, kernel arguments, and the thread id (if on
 // CPU).
 struct RuntimeContext {
+  enum class DevAllocType : int8_t { ndarray = 0, texture = 1, image = 2 };
```
Comment (k-ye) [inline taichi/program/context.h:17]: nit: Let's use `kNdarray` `kTexture` style, see https://google.github.io/styleguide/cppguide.html#Enumerator_Names

File: taichi/program/program.cpp [hunk 3]
```diff
@@ -564,6 +564,22 @@ Ndarray *Program::create_ndarray(const DataType type,
...
+    textures_.emplace_back(this, type, num_channels, shape[0], 1, 1);
+  } else if (shape.size() == 2) {
+    textures_.emplace_back(this, type, num_channels, shape[0], shape[1], 1);
+  } else if (shape.size() == 3) {
+    textures_.emplace_back(this, type, num_channels, shape[0], shape[1],
+                           shape[2]);
+  } else {
+    TI_ERROR("Texture shape invalid");
+  }
+  return &textures_.back();
```
Comment (k-ye) [inline taichi/program/program.cpp:580]: You don't have pointer stability here, consider using `std::vector<std::unique_ptr<Texture>>`

File: taichi/program/program.cpp [hunk 4]
```diff
@@ -564,6 +564,22 @@ Ndarray *Program::create_ndarray(const DataType type,
   return ndarrays_.back().get();
 }
 
+Texture *Program::create_texture(const DataType type,
+                                 int num_channels,
+                                 const std::vector<int> &shape) {
```
Comment (k-ye) [inline taichi/program/program.cpp:569]: Slightly nit: pass in a copy of `shape`, then pad it to length 3 with `1`? This saves a few if-else checks.

File: taichi/program/program.cpp (hunk 3 above)
Comment (bobcao3) [reply taichi/program/program.cpp:580]: ooo right I forgot about that, was wondering why the ndarrays uses a wrapped pointer

File: taichi/program/program.cpp (hunk 4 above)
Comment (bobcao3) [reply taichi/program/program.cpp:569]: This is intended to support 1D/2D/3D images, so there will always be if checks i think...

File: taichi/program/context.h [hunk 5]
```diff
@@ -60,15 +69,24 @@ struct RuntimeContext {
     this->array_runtime_sizes[i] = size;
   }
 
-  void set_array_is_device_allocation(int i, bool is_device_allocation) {
-    this->is_device_allocations[i] = is_device_allocation;
+  void set_array_is_device_allocation(
+      int i,
+      bool is_device_allocation,
```
Comment (k-ye) [inline taichi/program/context.h:74]: nit: We can remove this as well :-) (probably in a separate PR)

**Review by k-ye — APPROVED** [review]
LGTM!

File: taichi/program/texture.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+namespace lang {
+
+class Program;
+class Ndarray;
+
+class TI_DLL_EXPORT Texture {
+ public:
+  /* Constructs a Ndarray managed by Program.
+   * Memory allocation and deallocation is handled by Program.
+   * TODO: Ideally Ndarray shouldn't worry about memory alloc/dealloc at all.
```
Comment (ailzhang) [inline taichi/program/texture.h:20]: nit: update the comments to match texture?

**Review by ailzhang — APPROVED** [review]
LGTM!

**Pull request merged**